### PR TITLE
JENKINS-64601 Switch to github-api auto-refresh auth tokens

### PR DESCRIPTION
**Pull request by timja**
# Description

Auto refresh github app tokens via new API in github-api. See 
[JENKINS-64601](https://issues.jenkins-ci.org/browse/JENKINS-64601) for further information. 

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Automated tests have been added to exercise the changes
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verify that the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

# Documentation changes
- [ ] Link to jenkins.io PR, or an explanation for why no doc changes are needed

# Users/aliases to notify

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java [hunk 1]
```diff
@@ -122,11 +126,25 @@ public void setOwner(String owner) {
...
+    public AuthorizationProvider getAuthorizationProvider() {
+        // TODO create sub class
+        AuthorizationProvider authorizationProvider;
+        try {
+            authorizationProvider = new JWTTokenProvider(appID, privateKey.getPlainText());
+        } catch (GeneralSecurityException e) {
+            throw new IllegalArgumentException("Couldn't parse private key for GitHub app, make sure it's PKCS#8 format", e);
+        }
+
+        return new OrgAppInstallationAuthorizationProvider(owner, authorizationProvider);
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java:139]: this auth provider doesn't meet our needs and will need customising but it's enough for initial testing

Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java:139]: @timja 
Yeah, I figured this would be the case. We'll need an authorization provider that takes the credentials.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 2]
```diff
@@ -340,15 +340,21 @@ public static void checkApiUrlValidity(@Nonnull GitHub gitHub, @CheckForNull Sta
...
         if (credentials == null) {
             username = null;
             password = null;
             hash = "anonymous";
             authHash = "anonymous";
+        } else if (credentials instanceof GitHubAppCredentials) {
+            gitHubAppCredentials = (GitHubAppCredentials) credentials;
+            hash = Util.getDigestOf(gitHubAppCredentials.getAppID() + SALT); // want to ensure pooling by credential
+            authHash = Util.getDigestOf(gitHubAppCredentials.getAppID() + "::" + jenkins.getLegacyInstanceId());
+            username = gitHubAppCredentials.getUsername();
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:357]: This is an additional enhancement.  Perhaps it could go in a separate PR?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 3]
```diff
@@ -369,8 +375,12 @@ public static void checkApiUrlValidity(@Nonnull GitHub gitHub, @CheckForNull Sta
...
                 GitHubBuilder gb = createGitHubBuilder(apiUrl, cache);
 
-                if (username != null) {
-                    gb.withPassword(username, password);
+                if (username != null && password != null) {
+                    gb.withAuthorizationProvider(ImmutableAuthorizationProvider.fromOauthToken(password, username));
+                }
+
+                if (gitHubAppCredentials != null) {
+                    gb.withAuthorizationProvider(gitHubAppCredentials.getAuthorizationProvider());
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:383]:
```suggestion
                if (username != null) {
                    gb.withPassword(username, password);
                } 

                // For github app, use the auto-refresh provider
                if (credentials instanceof GitHubAppCredentials) {
                    gb.withAuthorizationProvider(((GitHubAppCredentials) credentials).getAuthorizationProvider());
```

Comment (timja) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:383]: without adding the `&& password != null` this will call withPassword when and authorization provider is set for github app, not sure if we want that

**Comment by timja** [comment]
Probably could do with some refactoring and tests but it seems to work now.

not sure if we want to re-write the `generateAppInstallationToken()` part of `GitHubAppCredentials` as well

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 2 above)
Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Okay, I like change you made to use all the fields.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 3 above)
Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:383]: That is what we currently do and it works.

**Comment by bitwiseman** [comment]
@timja 
Looks like we both implemented a solution.  
I made mine a PR against this branch: https://github.com/timja/github-branch-source-plugin/pull/1

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 4]
```diff
@@ -369,10 +375,15 @@ public static void checkApiUrlValidity(@Nonnull GitHub gitHub, @CheckForNull Sta
 
                 GitHubBuilder gb = createGitHubBuilder(apiUrl, cache);
 
-                if (username != null) {
-                    gb.withPassword(username, password);
+                if (gitHubAppCredentials != null) {
+                    gb.withAuthorizationProvider(gitHubAppCredentials.getAuthorizationProvider());
+                } else if (username != null && password != null) {
+                    // At the time of this change this works for OAuth tokens as well.
+                    // This may not continue to work in the future, as GitHub has deprecated Login/Password credentials.
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:382]: at some point should really switch to SecretText credentials so you can easily tell the difference...

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java [hunk 5]
```diff
@@ -48,7 +50,7 @@
     private static final String ERROR_AUTHENTICATING_GITHUB_APP = "Couldn't authenticate with GitHub app ID %s";
     private static final String NOT_INSTALLED = ", has it been installed to your GitHub organisation / user?";
 
-    private static final String ERROR_NOT_INSTALLED = ERROR_AUTHENTICATING_GITHUB_APP + NOT_INSTALLED;
+    public static final String ERROR_NOT_INSTALLED = ERROR_AUTHENTICATING_GITHUB_APP + NOT_INSTALLED;
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java:53]:
```suggestion
    private static final String ERROR_NOT_INSTALLED = ERROR_AUTHENTICATING_GITHUB_APP + NOT_INSTALLED;
```

**Comment by timja** [comment]
A backported version of this PR is now running on ci.jenkins.io and I was able to complete an org scan 🎉 

```
09:37:45 2115 repositories were processed
[Sat Jan 16 09:37:45 UTC 2021] Finished organization scan. Scan took 1 hr 6 min
Evaluating orphaned items in Plugins
Will not remove gradle-jpi-plugin as it is only #1 in the list
Will not remove gradle-jpi-plugin because it is new
Finished: SUCCESS
```

**Comment by bitwiseman** [comment]
@timja 
Added a test that verifies the behavior at least on the controller. I'm not sure how an AuthenticationProvider would end up on an agent.

**Comment by timja** [comment]
> @timja
> Added a test that verifies the behavior at least on the controller. I'm not sure how an AuthenticationProvider would end up on an agent.

Neither, I would have expected to hear shouting about ci.jenkins.io if it was a problem, and I haven't seen any issues

but that test is failing for some reason

**Comment by bitwiseman** [comment]
@timja Passed this time.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java [hunk 6]
```diff
@@ -209,6 +244,66 @@ public void setUpWireMock() throws Exception {
                             "}")));
     }
 
+    @Test
+    public void testProviderRefresh() throws Exception {
+        long notStaleSeconds = GitHubAppCredentials.AppInstallationToken.NOT_STALE_MINIMUM_SECONDS;
+        try {
+            appCredentials.setApiUri(githubApi.baseUrl());
+
+            // We want to demonstrate successful caching without waiting for a the default 1 minute
```
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java:253]:
```suggestion
            // We want to demonstrate successful caching without waiting for the default 1 minute
```

**Review by dwnusbaum — APPROVED** [review]
Looks like this is building on top of https://github.com/hub4j/github-api/pull/945, and the main change is the use of the new `AuthorizationProvider` interface. It is getting tricky for me to follow the different levels of credential caching and refreshing in this plugin, but I don't see anything obviously wrong, and if the PR has been tested and fixes the issue on `ci.jenkins.io` then it seems fine to me.
